x/txfees/keeper/fees.go

```diff
@@ -152,19 +150,27 @@ func (k Keeper) swapFeeToBaseDenom(
 	}
 
 	// Swap the coin to base denom
-	var (
-		tokenOutAmount = sdk.ZeroInt() // Token amount in base denom
-		route          = []poolmanagertypes.SwapAmountInRoute{{
-			PoolId:        feetoken.PoolID,
-			TokenOutDenom: baseDenom,
-		}}
-	)
-	err = osmoutils.ApplyFuncIfNoError(ctx, func(ctx sdk.Context) error {
-		tokenOutAmount, err = k.poolManager.RouteExactAmountIn(ctx, moduleAddr, route, takerFeeCoin, sdk.ZeroInt())
-		return err
-	})
+	route := []poolmanagertypes.SwapAmountInRoute{{
+		PoolId:        feetoken.PoolID,
+		TokenOutDenom: baseDenom,
+	}}
+
+	cacheCtx, write := osmoutils.NoEventCacheContext(ctx)
+	tokenOutAmount, err := k.poolManager.RouteExactAmountIn(cacheCtx, moduleAddr, route, takerFeeCoin, sdk.ZeroInt())
 	if err != nil {
-		return nil, sdk.Coins{takerFeeCoin}, nil
+		return nil, nil, fmt.Errorf("swap fee token: %w", err)
+	}
+	write()
+
+	// if we manage, we modify the last event to include the taker fee attribute
+	emittedEvents := cacheCtx.EventManager().Events()
+	if len(emittedEvents) > 0 && emittedEvents[len(emittedEvents)-1].Type == gammtypes.TypeEvtTokenSwapped {
+		ev := emittedEvents[len(emittedEvents)-1].AppendAttributes(sdk.NewAttribute(AttributeKeyTakerFee, "true"))
+
+		emittedEvents = emittedEvents[:len(emittedEvents)-1]
+		emittedEvents = append(emittedEvents, ev)
+
+		ctx.EventManager().EmitEvents(emittedEvents)
 	}
```

Comment:
suggestion to be put in a seperate function and clean up the naming and array manipulation 

```go 
func updateLastEventWithTakerFeeAttribute(ctx sdk.Context, cacheCtx sdk.Context) {
    events := cacheCtx.EventManager().Events()
    if len(events) == 0 {
        return
    }

    lastEvent := events[len(events)-1]
    if lastEvent.Type != gammtypes.TypeEvtTokenSwapped {
        return
    }

    // Add taker fee attribute to the last event
    updatedEvent := lastEvent.AppendAttributes(sdk.NewAttribute(AttributeKeyTakerFee, "true"))
    
    // Replace last event with updated version
    events[len(events)-1] = updatedEvent
    ctx.EventManager().EmitEvents(events)
}
```